[Core] Update OSUtilities

loumalouomega opened the pull request:
This updates the OSUtilities after the commentaries from this morning. It also adds new features (RemoveOnCurrentWorkingDir, IsDirExist, CreateDir) and the corresponding tests. I have compiled on Windows

I updated the vtk_output to create a folder when this doesn't exist

loumalouomega committed:
Adding new utilities to OSUtilities

loumalouomega committed:
Adding new tests

loumalouomega committed:
Update test

loumalouomega committed:
Improve vtk_output

loumalouomega committed:
Compiling on Windows

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

loumalouomega committed:
On Windows folder and files are different

loumalouomega committed:
Update test

philbucher commented on kratos/input_output/vtk_output.cpp:
```diff
@@ -159,16 +160,26 @@ std::string VtkOutput::GetOutputFileName(const ModelPart& rModelPart, const bool
 
     // Putting everything together
+    const std::string& r_folder_name = mOutputSettings["folder_name"].GetString();
+    const bool directory_exists = OSUtilities::IsDirExist(r_folder_name);
```
I am not overly happy with this (because so far we always did this in python) but I guess this is the only way if we are moving towards using this also in C (I assume you are doing this aren't you)?

Also if we have OS-independent solutions now I think it is ok

loumalouomega replied:
I am moving between Linux and Windows, now in Windows works, and in Linux looks like it fails. Second iteration

philbucher commented on kratos/tests/cpp_tests/utilities/test_os_utilities.cpp:
```diff
@@ -31,5 +34,36 @@ KRATOS_TEST_CASE_IN_SUITE(GetCurrentWorkingDir, KratosCoreFastSuite)
+KRATOS_TEST_CASE_IN_SUITE(SeveralUtilities, KratosCoreFastSuite)
+{
+    // Create dir
+    const std::string name = "AUXILIAR_RIDICULOUS_NAME";
```
```suggestion
    const std::string name = "auxiliar_ridiculous_name";
```

loumalouomega replied:
Ok

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -10,10 +10,15 @@
 //
 
+// Project includes
+#include "utilities/os_utilities.h"
```
```suggestion
#include "utilities/os_utilities.h" # has to be included before using KRATOS_COMPILED_IN_WINDOWS
```

to prevent this from being moved in the future

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -41,5 +43,42 @@ std::string GetCurrentWorkingDir()
+/***********************************************************************************/
+/***********************************************************************************/
+
+bool IsDirExist(const std::string& rFolderName)
```
```suggestion
bool DirExists(const std::string& rFolderName)
```

~~~c
if (DirExists(...)) { ... }
~~~

is easier to understand I think

loumalouomega replied:
If I change this it will be changing all, otherwise many conflicts

philbucher commented on kratos/utilities/os_utilities.h:
```diff
@@ -55,6 +55,26 @@ namespace OSUtilities
+     * @brief This method removes a file on the current working directory
+     * @param rFileName The name of the file to be removed
+     */
+    void KRATOS_API(KRATOS_CORE) RemoveOnCurrentWorkingDir(const std::string& rFileName);
```
how abt two functions?
- `RemoveFile`
- `RemoveFileInCurrentWorkingDir`

loumalouomega replied:
OK

loumalouomega committed:
Comment

Co-Authored-By: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
snake_case

Co-Authored-By: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
C comment

loumalouomega committed:
Many fixes

loumalouomega commented:
OK, i go back to windows to check that I haven't broke anything :P

loumalouomega commented:
It does compile

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -41,5 +43,82 @@ std::string GetCurrentWorkingDir()
+/***********************************************************************************/
+/***********************************************************************************/
+
+bool DirExist(const std::string& rFolderName)
```
```suggestion
bool IsDir(const std::string& rFolderName)
```

loumalouomega replied:
but, but you told me...

philbucher replied:
I know, sorry, was not thinking :/

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -41,5 +43,82 @@ std::string GetCurrentWorkingDir()
+/***********************************************************************************/
+/***********************************************************************************/
+
+bool FileExist(const std::string& rFileName)
```
```suggestion
bool IsFile(const std::string& rFileName)
```

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -41,5 +43,82 @@ std::string GetCurrentWorkingDir()
+/***********************************************************************************/
+/***********************************************************************************/
+
+void RemoveOnCurrentWorkingDir(const std::string& rFileName)
```
```suggestion
void RemoveFileInCurrentWorkingDir(const std::string& rFileName)
```

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -41,5 +43,82 @@ std::string GetCurrentWorkingDir()
+/***********************************************************************************/
+/***********************************************************************************/
+
+void Remove(const std::string& rFileName)
```
do you want this to remove a file or a folder?

=> maybe separate fcts are better here?
`RemoveFile` & `RemoveDir`

loumalouomega replied:
ok

loumalouomega committed:
Improve test + trigger build

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -41,5 +43,82 @@ std::string GetCurrentWorkingDir()
+/***********************************************************************************/
+/***********************************************************************************/
+
+std::string AppendFolderFile(
```
this sounds strange
```suggestion
std::string JoinPath(
```

loumalouomega replied:
ok

philbucher reviewed (commented):
some more comments, sorry for confusion, now I see how you use it
My suggestions aim to have same/similar names as in python

loumalouomega committed:
Missing COTIRE macros

loumalouomega committed:
Several suggestions

loumalouomega committed:
Adding ;

loumalouomega committed:
Revert "Adding ;"

This reverts commit b4b7d4d2cc596dabccea7f3dd86e215d98793021.

loumalouomega committed:
Revert "Missing includes that fail on COTIRE Appveyor for some reason"

This reverts commit 82c89848d39a92928b0eb21490e9e05edd5af1b0.

loumalouomega committed:
Missing includes that fail on COTIRE Appveyor for some reason

loumalouomega committed:
Missing includes that fail on COTIRE Appveyor for some reason

loumalouomega committed:
Revert "Missing includes that fail on COTIRE Appveyor for some reason"

This reverts commit 690bea16ab74c1c0d0a3816c76b7eb4f6d9c9889.

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -41,5 +49,112 @@ std::string GetCurrentWorkingDir()
+void RemoveFile(const std::string& rFileName)
+{
+    const char* name = rFileName.c_str();
+    remove(name);
```
what happens if this is not a file?
Would it make sense to use `IsFile` here?

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -41,5 +49,112 @@ std::string GetCurrentWorkingDir()
+{
+    const char* name = rFolderName.c_str();
+#ifdef KRATOS_COMPILED_IN_WINDOWS
+    RemoveDirectoryA(name);
```
same

philbucher reviewed (commented):
almost ready :+1:

loumalouomega committed:
Last suggestions

loumalouomega committed:
Hopefully solving problem

loumalouomega committed:
Adding warning

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -41,5 +44,132 @@ std::string GetCurrentWorkingDir()
+        const char* name = rFileName.c_str();
+        remove(name);
+    } else {
+        KRATOS_WARNING("OSUtilities") << "This is a folder. Please use RemoveDir" << std::endl;
```
```suggestion
        KRATOS_WARNING("OSUtilities") << "\"" << rFileName << "\" could not be removed because it is not a file" << std::endl;
```

philbucher replied:
same

philbucher commented on kratos/utilities/os_utilities.cpp:
```diff
@@ -41,5 +44,132 @@ std::string GetCurrentWorkingDir()
+        rmdir(name);
+#endif
+    } else {
+        KRATOS_WARNING("OSUtilities") << "This is a file. Please use RemoveFile" << std::endl;
```
```suggestion
        KRATOS_WARNING("OSUtilities") << "\"" << rFolderName << "\" could not be removed because it is not a folder" << std::endl;
```

philbucher replied:
you cannot say that it is a file, could also be a link or sth else

philbucher commented:
hm appveyor still fails

loumalouomega committed:
Suggestion (I)

Co-Authored-By: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega committed:
Suggestion (II)

Co-Authored-By: Philipp Bucher <philipp.bucher@tum.de>

loumalouomega commented:
> hm appveyor still fails

Yes, and I don't know why, in my machine works. @roigcarlo ?,

~~~
  C:\projects\kratos\kratos\modeler\tetrahedra_ball.cpp(43): error C2589: '(': illegal token on right side of '::' (compiling source file C:\projects\kratos\cmake_build\kratos\Debug\cotire\KratosCore_CXX_unity_0_98.cxx) [C:\projects\kratos\cmake_build\kratos\KratosCore_unity.vcxproj]
  C:\projects\kratos\kratos\modeler\tetrahedra_ball.cpp(43): error C2062: type 'unknown-type' unexpected (compiling source file C:\projects\kratos\cmake_build\kratos\Debug\cotire\KratosCore_CXX_unity_0_98.cxx) [C:\projects\kratos\cmake_build\kratos\KratosCore_unity.vcxproj]
  C:\projects\kratos\kratos\modeler\tetrahedra_ball.cpp(43): error C2059: syntax error: ')' (compiling source file C:\projects\kratos\cmake_build\kratos\Debug\cotire\KratosCore_CXX_unity_0_98.cxx) [C:\projects\kratos\cmake_build\kratos\KratosCore_unity.vcxproj]
  C:\projects\kratos\kratos\modeler\tetrahedra_ball.cpp(45): error C2589: '(': illegal token on right side of '::' (compiling source file C:\projects\kratos\cmake_build\kratos\Debug\cotire\KratosCore_CXX_unity_0_98.cxx) [C:\projects\kratos\cmake_build\kratos\KratosCore_unity.vcxproj]
  C:\projects\kratos\kratos\modeler\tetrahedra_ball.cpp(45): error C2062: type 'unknown-type' unexpected (compiling source file C:\projects\kratos\cmake_build\kratos\Debug\cotire\KratosCore_CXX_unity_0_98.cxx) [C:\projects\kratos\cmake_build\kratos\KratosCore_unity.vcxproj]
  C:\projects\kratos\kratos\modeler\tetrahedra_ball.cpp(45): error C2059: syntax error: ')' (compiling source file C:\projects\kratos\cmake_build\kratos\Debug\cotire\KratosCore_CXX_unity_0_98.cxx) [C:\projects\kratos\cmake_build\kratos\KratosCore_unity.vcxproj]
~~~

Not even related with these files

philbucher commented:
a test fails in Travis

philbucher commented:
also codacy has some strange complaints

loumalouomega committed:
Revert

loumalouomega committed:
Removing check

loumalouomega commented:
If someone that uses Windows can help with this will be very helpful

philbucher commented:
ping @tteschemacher @AlejandroCornejo

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

loumalouomega committed:
Revert "[Core] Hash and comparator not properly defined"

This reverts commit 80d59514ea6b192c88962a433c3879c7c9f655d0.

loumalouomega committed:
[Core] Hash and comparator not properly defined

This give me many warnings on the last version of clang (9). I moved it to the key_hash file, so everyone can use it

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

philbucher commented:
ping @tteschemacher

loumalouomega commented:
I am not able to fix the remaining fails in Windows

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

philbucher commented:
I need the same thing for sth else and had again a look at this => ping @tteschemacher ;)

I also checked out boost::filesystem but it is not header only :(((

filesystem is only included in c++17 :/

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

loumalouomega committed:
Minor

loumalouomega commented:
@philbucher In my Windows computer it works, I don't know why not here

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

loumalouomega committed:
Merge branch 'master' into core/os-utilities-improve

philbucher commented:
Since this is a limited interface to `std::filesystem` (part of c++17), we could think about using [this](https://github.com/gulrak/filesystem) instead
It is basically an `std::filesystem` implementation for c++11, working in the same style as pybind (header-only etc...)

I will report soon my experiences

philbucher commented:
#6061
